### DISCUSSION TITLE
Renovate doesn't seem to pick up private package feeds defined via main config

### DISCUSSION THREAD

**Discussion by horihel** [renovatebot/renovate#13589] [Request Help]
# Config
- Renovate 31.28.1
- Repository does not contain nuget.config files
- config.js contains a nuget configuration snippet:
    ```js
    nuget: {
      registryUrls: [
        "https://api.nuget.org/v3/index.json",
        "https://pkgs.dev.azure.com/myorg/_packaging/myfeed-public%40Release/nuget/v3/index.json",
        "https://pkgs.dev.azure.com/myorg/project/_packaging/myfeed2%40Release/nuget/v3/index.json",
      ],
    },
    ```
- it also contains the corresponding auth hostRules
    ```js
      hostRules: [
      {
        hostType: "nuget",
        matchHost: "pkgs.dev.azure.com",
        username: "apikey",
        password: process.env.TOKEN,
      },
    ],
    ```
# Expectation
Renovate uses a generated nuget.config with the 3 repositories for all nuget related tasks, including authentication

# What seems to happen
Renovate only adds the public nuget repo, completely ignoring the other two and then restore fails

i'm seeing only "dotnet nuget add source" lines in the log for the public repo, the private repos are not mentioned in the logs at all (except in the very first line that summarizes the config object). It looks like my nuget: section is completely ignored.

Am I missing something?

**Comment by rarkins** [comment]
What leads you to expect that Renovate will generate a nuget config file?

**Reply by horihel** [reply]
The documentation does https://docs.renovatebot.com/nuget/#Alternate%20feeds
> Alternative feeds can be specified either in a NuGet.config file within your repository (Renovate will not search outside the repository) or in Renovate configuration options:

maybe I understood it wrong, but as far is I see, the only way of providing those configured feeds to the dependency process would be writing/appending to a nuget.config, wouldn't it?

**Reply by horihel** [reply]
renovate _does_ add the public repo by calling "nuget add source" - but just the public nuget repo, not the private ones

**Reply by rarkins** [reply]
Without such a file, renovate lookups succeed (it detects new versions) but running nuget to update lock files will likely fail. 

You're welcome to submit a detailed feature request if you can find the relevant nuget docs on nuget file formats including mandatory fields, whether there's differences between nuget versions, etc. 

Does anything prevent you from adding the file to your repo?

**Reply by horihel** [reply]
doesn't it already create such a file (by running "nuget add source")?
I just don't understand why it only runs "nuget add source" once (for https://api.nuget.org/v3/index.json) instead of 3 times (for all configured repos).

this log line
```json
{
	"name": "renovate",
	"hostname": "renovatebot-manual-v6f-fvnqf",
	"pid": 8,
	"level": 20,
	"logContext": "qd_74HPiLZERfwQwB2pw8",
	"repository": "myproject/myrepo",
	"branch": "renovate/google.protobuf-3.x",
	"cmd": [
		"dotnet nuget add source https://api.nuget.org/v3/index.json --configfile /tmp/renovate/cache/others/nuget/5bb40432134e189d/nuget.config",
		"dotnet restore ods/OdsService/OdsService.csproj --force-evaluate --configfile /tmp/renovate/cache/others/nuget/5bb40432134e189d/nuget.config"
	],
	"msg": "dotnet command",
	"time": "2022-01-17T10:03:54.089Z",
	"v": 0
}
```

should contain a nuget add source for all repos, shouldn't it?

**Reply by horihel** [reply]
we opted for nuking all nuget.config files from the repos because developers can use unpromoted nuget packages during development (in Azure the @Local feed) - during pipeline runs, only packages from the @Release feed can be used (actually it's a bit flexible, depending on the branch). This workflow is suggested by Azure Artifacts, promoting packages to a different view once they're considered fit for publishing.
As Azure Pipelines supports replacing nuget config during build - and developers just used global feeds in Visual Studio anyweays, we opted to remove them rather than having them rot unused.
I hoped I could dodge the bullet of having to place them all back in the repos again. There's quite a few of them.

**Reply by rarkins** [reply]
OK, thanks for that feedback. Do you know if it's possible to avoid Renovate needing to create the `nuget.config` too, as long as it runs the appropriate CLI commands?

**Reply by horihel** [reply]
I think the nuget.config file must always be created (and Renovate already creates an empty one [here](https://github.com/renovatebot/renovate/blob/2fd071cadf24e4c21a5255f25bc8b521f838d3ff/lib/manager/nuget/artifacts.ts#L73) )

the only piece missing (IMHO) is adding those additional repos via the CLI commands.

**Comment by viceice** [comment]
https://github.com/renovatebot/renovate/blob/2fd071cadf24e4c21a5255f25bc8b521f838d3ff/lib/manager/nuget/artifacts.ts#L30-L58

This is the code to add registries to artifact update

**Reply by horihel** [reply]
does getConfiguredRegistries consider sources specified in config.js?

**Reply by viceice** [reply]
Nope, currently it only reads them from nuget.config.

**Reply by horihel** [reply]
Then I don't understand what https://docs.renovatebot.com/nuget/#Alternate%20feeds is supposed to do or where it is used.

**Reply by viceice** [reply]
It was written before we could update lockfiles to find package updates in other registries.

But sure, we should extend the option to add all  registry urls from renovate config to the nuget config, if the file don't exist.

**Reply by horihel** [reply]
any way I can help with this? Should I write up a feature request referring to this?

**Reply by viceice** [reply]
Yes, a feature request, no bug report. 😉

**Comment by mnovikov-mindbox** [comment]
Hi! Was feature request / bug registered? Is this problem still exists?

**Reply by horihel** [reply]
it was discussed in #13591
